mdstream/mdstream.go

```diff
@@ -68,47 +73,137 @@ func DecodeProposalGeneral(payload []byte) (*ProposalGeneral, error) {
 //
 // This mdstream is used by both pi and cms.
-// XXX this is missing the admin signature
-type RecordStatusChange struct {
+type RecordStatusChangeV1 struct {
 	Version             uint             `json:"version"`                       // Version of the struct
 	AdminPubKey         string           `json:"adminpubkey"`                   // Identity of the administrator
 	NewStatus           pd.RecordStatusT `json:"newstatus"`                     // New status
-	StatusChangeMessage string           `json:"statuschangemessage,omitempty"` // Status change message
-	Timestamp           int64            `json:"timestamp"`                     // Timestamp of the change
+	StatusChangeMessage string           `json:"statuschangemessage,omitempty"` // Change message
+	Timestamp           int64            `json:"timestamp"`                     // UNIX timestamp
 }
 
-// EncodeRecordStatusChange encodes an RecordStatusChange into a JSON byte
+// EncodeRecordStatusChangeV1 encodes an RecordStatusChangeV1 into a JSON byte
 // slice.
-func EncodeRecordStatusChange(rsc RecordStatusChange) ([]byte, error) {
+func EncodeRecordStatusChangeV1(rsc RecordStatusChangeV1) ([]byte, error) {
 	b, err := json.Marshal(rsc)
 	if err != nil {
 		return nil, err
 	}
 	return b, nil
 }
 
-// DecodeRecordStatusChange decodes a JSON byte slice into a slice of
-// RecordStatusChange.
-func DecodeRecordStatusChange(payload []byte) ([]RecordStatusChange, error) {
-	var changes []RecordStatusChange
+// RecordStatusChangeV2 represents a politeiad record status change and is used
+// to store additional status change metadata that would not otherwise be
+// captured by the politeiad status change routes.
+//
+// V2 adds the Signature field, which was erroneously left out of V1.
+//
+// This mdstream is used by both pi and cms.
+type RecordStatusChangeV2 struct {
+	Version             uint             `json:"version"`                       // Struct version
+	NewStatus           pd.RecordStatusT `json:"newstatus"`                     // New status
+	StatusChangeMessage string           `json:"statuschangemessage,omitempty"` // Change message
+	Signature           string           `json:"signature"`                     // Signature of (Token + NewStatus + StatusChangeMessage)
+	AdminPubKey         string           `json:"adminpubkey"`                   // Signature pubkey
+	Timestamp           int64            `json:"timestamp"`                     // UNIX timestamp
+}
+
+// VerifySignature verifies that the RecordStatusChangeV2 signature is correct.
+func (r *RecordStatusChangeV2) VerifySignature(token string) error {
+	sig, err := util.ConvertSignature(r.Signature)
+	if err != nil {
+		return err
+	}
+	b, err := hex.DecodeString(r.AdminPubKey)
+	if err != nil {
+		return err
+	}
+	pk, err := identity.PublicIdentityFromBytes(b)
+	if err != nil {
+		return err
+	}
+	msg := token + strconv.Itoa(int(r.NewStatus)) + r.StatusChangeMessage
```

Comment:
Should we add the timestamp here?

Reply:
We don't include the timestamp because its set by politeiawww, not by the client.  Adding it to the signature would require API changes.